### PR TITLE
Use application cache to reuse patient sets.

### PR DESCRIPTION
**Pull request by forus**
- Don't use database field for that. Doesn't work for oracle as it has clob that you could not apply indexes on. Use the application cache instead.
- Don't reuse query master objects at all.

### CODE REVIEW

**Comment by forus** [comment]
The patient set table could grow fast that could cause the disk to run out of space. But it's separate concern anyway. TODO Create a ticket to take care of this when PR succeed.

**Comment by forus** [comment]
@ewelinagr so far it fixes TMT-378, but not TMT-379 yet. The plan is to implement TMT-379 fix as separate PR or commit.

**Comment by forus** [comment]
I should use the cache library instead.

File: transmart-core-db/grails-app/services/org/transmartproject/db/clinical/PatientSetService.groovy [hunk 1]
```diff
@@ -716,16 +731,46 @@ class PatientSetService extends AbstractDataResourceService implements PatientSe
...
+     * Clear  query result ids cache.
+     * This function should be called after data loading.
      */
-    @Transactional
-    void clearPatientSets() {
-        log.info 'Clearing patient sets ...'
-        QtPatientSetCollection.executeUpdate('delete QtPatientSetCollection')
-        QtQueryResultInstance.executeUpdate('delete QtQueryResultInstance')
-        QtQueryInstance.executeUpdate('delete QtQueryInstance')
+    void clearPatientSetIdsCache() {
```
Comment (ewelinagr) [inline transmart-core-db/grails-app/services/org/transmartproject/db/clinical/PatientSetService.groovy]: Might be useful to change the documentation. https://transmart-test.thehyve.net/open-api/index.html#!/v2/get_v2_admin_system_after_data_loading_update mentions clearing patient sets, so can be confusing

File: transmart-core-db/grails-app/services/org/transmartproject/db/clinical/PatientSetService.groovy [hunk 2]
```diff
@@ -716,16 +731,46 @@ class PatientSetService extends AbstractDataResourceService implements PatientSe
...
+        log.info 'Clearing patient set ids cache ...'
+        getQueryResultIdsCache().clear()
+    }
+
+    /**
+     * Get's query result id (aka patient set id) for the given user and constraint from the cache.
+     * The cache meant to be cleaned after the data loading. {@see clearPatientSetIdsCache}
```
Comment (ewelinagr) [inline transmart-core-db/grails-app/services/org/transmartproject/db/clinical/PatientSetService.groovy]: @see -> @link

**Comment by forus** [comment]
@ewelinagr FYI I've created a ticket to implement a job to remove "old" patient sets periodically https://jira.thehyve.nl/browse/TMT-395. I also started using `Contraint.toJson()` for now. There is a ticket to focus on improving the performance of calculating cache keys out of the constraints. https://jira.thehyve.nl/browse/TMT-395

File: transmart-core-db/grails-app/services/org/transmartproject/db/clinical/PatientSetService.groovy [hunk 3]
```diff
@@ -152,27 +129,65 @@ class PatientSetService extends AbstractDataResourceService implements PatientSe
...
-                .add(Restrictions.eq('qi.userId', user.username))
-                .add(Restrictions.eq('qi.deleteFlag', 'N'))
-                .add(Restrictions.eq('qm.requestConstraints', constraint.toJson()))
-                .add(Restrictions.eq('qm.deleteFlag', 'N'))
-                .addOrder(Order.desc('qri.endDate'))
-        (QueryResult)getFirst(criteria)
+    QueryResult findFinishedQueryResultInCacheBy(User user,
+                                                 Constraint constraint) {
+        Long queryResultId = getQueryResultIdFromCache(user, constraint)
+        if (log.debugEnabled) {
```
Comment (gijskant) [inline transmart-core-db/grails-app/services/org/transmartproject/db/clinical/PatientSetService.groovy]: Why check?

File: transmart-core-db/grails-app/services/org/transmartproject/db/clinical/PatientSetService.groovy [hunk 4]
```diff
@@ -152,27 +129,65 @@ class PatientSetService extends AbstractDataResourceService implements PatientSe
...
+            log.debug("Get query result ${queryResultId} for ${user} and ${constraint}")
+        }
+        if (queryResultId) {
+            QtQueryResultInstance queryResult = QtQueryResultInstance.get(queryResultId)
+            boolean usable = queryResult && queryResult.status == QueryStatus.FINISHED && queryResult.deleteFlag == 'N'
```
Comment (gijskant) [inline transmart-core-db/grails-app/services/org/transmartproject/db/clinical/PatientSetService.groovy]: If you reverse the order, the massive indenting is not required and also a separate function to check this is not needed. E.g.,:
```groovy
if (!queryResultId) {
  return null
}
...
if (!queryResult || queryResult.status != FINISHED) {
  return null
}
```

File: transmart-core-db/grails-app/services/org/transmartproject/db/clinical/PatientSetService.groovy [hunk 5]
```diff
@@ -152,27 +129,65 @@ class PatientSetService extends AbstractDataResourceService implements PatientSe
...
-                .add(Restrictions.eq('qri.queryResultType', patientSetResultType))
-                .add(Restrictions.eq('qri.deleteFlag', 'N'))
-                .add(Restrictions.eq('qri.statusTypeId', (short)QueryStatus.FINISHED.id))
-                .add(Restrictions.eq('qi.userId', user.username))
-                .add(Restrictions.eq('qi.deleteFlag', 'N'))
-                .add(Restrictions.eq('qm.requestConstraints', constraint.toJson()))
-                .add(Restrictions.eq('qm.deleteFlag', 'N'))
-                .addOrder(Order.desc('qri.endDate'))
-        (QueryResult)getFirst(criteria)
+    QueryResult findFinishedQueryResultInCacheBy(User user,
```
Comment (gijskant) [inline transmart-core-db/grails-app/services/org/transmartproject/db/clinical/PatientSetService.groovy]: Consequence is that the persisted patient sets are useless after restarting the service, right?

File: transmart-data/updatedb/release-17.1-HYVE-5/postgres/migrate_bitset.sql [hunk 6]
```diff
@@ -1,2 +1,4 @@
 \i ../../../ddl/postgres/biomart_user/views/patient_num_boundaries.sql
-\i ../../../ddl/postgres/biomart_user/views/patient_set_bitset.sql
\ No newline at end of file
+\i ../../../ddl/postgres/biomart_user/views/patient_set_bitset.sql
+
+drop index i2b2demodata.qt_query_master_request_constraints;
```
Comment (gijskant) [inline transmart-data/updatedb/release-17.1-HYVE-5/postgres/migrate_bitset.sql]: Not for Oracle?

**Pull request merged**